compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/QueryDesugar.java

```diff
@@ -298,6 +311,38 @@ BLangStatementExpression desugar(BLangQueryAction queryAction, SymbolEnv env) {
         return stmtExpr;
     }
 
+    /**
+     * Get result type of the query output.
+     *
+     * @param type type of query expression.
+     * @return result type.
+     */
+    private BType resolveExprType(BType type) {
+        if (type.tag == TypeTags.STREAM) {
+            return ((BStreamType) type).constraint;
+        } else if (type.tag == TypeTags.TABLE) {
+            return ((BTableType) type).constraint;
+        } else if (type.tag == TypeTags.ARRAY) {
+            return ((BArrayType) type).eType;
+        } else if (type.tag == TypeTags.UNION) {
+            List<BType> exprTypes = new ArrayList<>(((BUnionType) type).getMemberTypes());
+            for (BType returnType : exprTypes) {
+                if (returnType.tag == TypeTags.STREAM) {
+                    return ((BStreamType) returnType).constraint;
+                } else if (returnType.tag == TypeTags.TABLE) {
+                    return ((BTableType) returnType).constraint;
+                }  else if (returnType.tag == TypeTags.ARRAY) {
+                    return ((BArrayType) returnType).eType;
+                } else if (returnType.tag == TypeTags.STRING) {
+                    return returnType;
+                } else if (returnType.tag == TypeTags.XML) {
+                    return returnType;
+                }
+            }
+        }
+        return type;
+    }
```

Comment:
You've written a similar function in CodeAnalyzer as well (but without string, XML checks). Is there a way to refactor both into one. Maybe we can but in Types?